File: mobile/lib/services/freesound_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:flutter_dotenv/flutter_dotenv.dart';

class FreesoundService {
  static const String baseUrl = 'https://freesound.org/apiv2';
  static final String apiKey = dotenv.env['FREESOUND_API_KEY'] ?? '';

  final Map<String, String> categories = {
    'nature': 'peaceful nature sounds relaxing',
    'meditation': 'tibetan bowls meditation',
    'ambient': 'ambient peaceful',
    'binaural': 'binaural beats meditation',
    'all': ''
  };

  Future<List<Map<String, dynamic>>> searchSounds(String token, {
    String query = '',
    String category = 'all',
    int page = 1
  }) async {
    try {
      final searchQuery = categories[category] ?? '';
      final fullQuery = query.isNotEmpty ? '$searchQuery $query' : searchQuery;

      final response = await http.get(
        Uri.parse('$baseUrl/search/text/')
            .replace(queryParameters: {
          'query': fullQuery,
          'page': page.toString(),
          'page_size': '12',
          'fields': 'id,name,username,previews,images,duration,description,avg_rating,tags',
          'token': apiKey,
        }),
      );

      print('Response status: ${response.statusCode}');
      print('Response body: ${response.body}');

      if (response.statusCode == 200) {
        final Map<String, dynamic> responseData = json.decode(response.body);
        return List<Map<String, dynamic>>.from(responseData['results'] ?? []);
      } else {
        throw Exception('Failed to load sounds: ${response.statusCode}');
      }
    } catch (e) {
      print('Freesound service error: $e');
      throw Exception('Failed to fetch sounds: $e');
    }
  }

  Future<Map<String, dynamic>> getSoundDetails(String token, String soundId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/sounds/$soundId/')
            .replace(queryParameters: {
          'fields': 'id,name,username,previews,images,duration,description,avg_rating,tags',
          'token': apiKey,
        }),
      );

      if (response.statusCode == 200) {
        return json.decode(response.body);
      } else {
        throw Exception('Failed to get sound details');
      }
    } catch (e) {
      throw Exception('Failed to get sound details: $e');
    }
  }

  Future<String> getAudioUrl(String token, String soundId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/sounds/$soundId/')
            .replace(queryParameters: {
          'fields': 'previews',
          'token': apiKey,
        }),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data['previews']['preview-hq-mp3'];
      } else {
        throw Exception('Failed to get audio URL');
      }
    } catch (e) {
      throw Exception('Failed to get audio URL: $e');
    }
  }

  Future<List<Map<String, dynamic>>> getFavorites(String token, String userId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/users/$userId/bookmark_categories/')
            .replace(queryParameters: {
          'token': apiKey,
        }),
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = json.decode(response.body);
        return List<Map<String, dynamic>>.from(data);
      } else {
        throw Exception('Failed to get favorites');
      }
    } catch (e) {
      throw Exception('Failed to get favorites: $e');
    }
  }

  Future<void> toggleFavorite(String token, String userId, String soundId) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/users/$userId/bookmark_categories/')
            .replace(queryParameters: {
          'token': apiKey,
        }),
        headers: {
          'Content-Type': 'application/json',
        },
        body: json.encode({
          'sound_id': soundId,
        }),
      );

      if (response.statusCode != 200) {
        throw Exception('Failed to toggle favorite');
      }
    } catch (e) {
      throw Exception('Failed to toggle favorite: $e');
    }
  }
}

File: web/src/services/freesoundService.ts
import { Sound, SearchFilters, SearchResponse } from '../types/wellnessHub.types';
import { API_ENDPOINTS } from '../utils/api';

const API_KEY = import.meta.env.VITE_FREESOUND_API_KEY;
const BASE_URL = 'https://freesound.org/apiv2';

interface Categories {
  [key: string]: string;
}

const categories: Categories = {
  nature: 'peaceful nature sounds relaxing',
  meditation: 'tibetan bowls meditation',
  rain: 'gentle rain ambient',
  ocean: 'calming ocean waves',
  forest: 'peaceful forest ambience',
  space: 'deep space ambient',
  bowls: 'crystal singing bowls healing',
  binaural: 'binaural beats meditation',
  flute: 'native american flute',
};

/**
 * Search for sounds on Freesound API
 * @param category - Category to search for
 * @param page - Page number for pagination
 * @param filters - Optional filters for the search
 * @returns Search results with sounds
 */
export const searchSounds = async (
  category: string,
  page: number = 1,
  filters: SearchFilters = {}
): Promise<SearchResponse> => {
  try {
    const { duration, rating, tags } = filters;
    let query = categories[category] || category;

    if (tags) {
      query += ` ${tags}`;
    }

    // Parse duration filter
    let durationFilter = '';
    if (duration) {
      durationFilter = `&filter=duration:${duration}`;
    }

    // Parse rating filter
    let ratingFilter = '';
    if (rating) {
      ratingFilter = `&filter=avg_rating:[${rating} TO 5]`;
    }

    const response = await fetch(
      `${BASE_URL}/search/text/?query=${encodeURIComponent(query)}${durationFilter}${ratingFilter}&fields=id,name,username,previews,images,duration,description,avg_rating,tags&page_size=12&page=${page}&token=${API_KEY}`
    );

    if (!response.ok) throw new Error('Failed to fetch sounds');

    return await response.json();
  } catch (error) {
    console.error('Error fetching sounds:', error);
    throw error;
  }
};

/**
 * Get sound details by ID
 * @param soundId - ID of the sound to fetch
 * @returns Sound details
 */
export const getSound = async (soundId: number | string): Promise<Sound> => {
  try {
    const response = await fetch(`${BASE_URL}/sounds/${soundId}/?token=${API_KEY}`);

    if (!response.ok) throw new Error('Failed to fetch sound');

    return await response.json();
  } catch (error) {
    console.error('Error fetching sound:', error);
    throw error;
  }
};

/**
 * Get proxied audio URL to bypass CORS issues
 * @param originalUrl - Original audio URL
 * @returns Proxied audio URL as blob
 */
export const getProxiedAudioUrl = async (originalUrl: string): Promise<string> => {
  try {
    const response = await fetch(API_ENDPOINTS.FREESOUND_AUDIO_PROXY, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ url: originalUrl }),
    });

    if (!response.ok) throw new Error('Failed to get proxied audio URL');

    const blob = await response.blob();
    return URL.createObjectURL(blob);
  } catch (error) {
    console.error('Error getting proxied audio:', error);
    throw error;
  }
};
